fused-effects-time/src/Control/Effect/Time.hs

```diff
@@ -0,0 +1,31 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+module Control.Effect.Time
+( -- * Time effect
+  now
+, timeWith
+, Time(..)
+  -- * Re-exports
+, Algebra
+, Has
+, run
+) where
+
+import Control.Algebra
+import Data.Kind (Type)
+
+now :: Has (Time instant) sig m => m instant
+now = send Now
+{-# INLINE now #-}
+
+timeWith :: Has (Time instant) sig m => (instant -> instant -> delta) -> m a -> m (delta, a)
+timeWith with m = do
+  start <- now
+  a <- m
+  end <- now
+  let d = with start end
+  d `seq` pure (d, a)
+{-# INLINE timeWith #-}
```

Comment:
Describing a general operation like this allows us to avoid having to introduce some sort of abstraction to describe diffing instants to produce deltas. In particular, we abstract the _generation_ of instants, and not the instants themselves.